src/ImageSharp/Formats/ImageEncoder.cs

```diff
@@ -1,43 +1,94 @@
+        => this.EncodeWithSeekableStreamAsync(image, stream, cancellationToken);
 
-/// <summary>
-/// The base class for all image encoders that allow color palette generation via quantization.
-/// </summary>
-public abstract class QuantizingImageEncoder : ImageEncoder
-{
     /// <summary>
-    /// Gets the quantizer used to generate the color palette.
+    /// Encodes the image to the specified stream from the <see cref="Image{TPixel}" />.
     /// </summary>
-    public IQuantizer Quantizer { get; init; } = KnownQuantizers.Octree;
+    /// <remarks>
+    /// This method is designed to support the ImageSharp internal infrastructure and is not recommended for direct use.
+    /// </remarks>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    /// <param name="image">The <see cref="Image{TPixel}" /> to encode from.</param>
+    /// <param name="stream">The <see cref="Stream" /> to encode the image data to.</param>
+    /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+    protected abstract void Encode<TPixel>(Image<TPixel> image, Stream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>;
 
-    /// <summary>
-    /// Gets the <see cref="IPixelSamplingStrategy"/> used for quantization when building color palettes.
-    /// </summary>
-    public IPixelSamplingStrategy PixelSamplingStrategy { get; init; } = new DefaultPixelSamplingStrategy();
+    private void EncodeWithSeekableStream<TPixel>(Image<TPixel> image, Stream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        Configuration configuration = image.GetConfiguration();
+        if (stream.CanSeek)
+        {
+            this.Encode(image, stream, cancellationToken);
+        }
+        else
+        {
+            using ChunkedMemoryStream ms = new(configuration.MemoryAllocator);
+            this.Encode(image, stream, cancellationToken);
+            ms.Position = 0;
+            ms.CopyTo(stream, configuration.StreamProcessingBufferSize);
+        }
+    }
+
+    private async Task EncodeWithSeekableStreamAsync<TPixel>(Image<TPixel> image, Stream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        Configuration configuration = image.GetConfiguration();
+        if (stream.CanSeek)
+        {
+            await DoEncodeAsync(stream).ConfigureAwait(false);
+        }
+        else
+        {
+            using ChunkedMemoryStream ms = new(configuration.MemoryAllocator);
+            await DoEncodeAsync(ms);
+            ms.Position = 0;
+            await ms.CopyToAsync(stream, configuration.StreamProcessingBufferSize, cancellationToken)
+                    .ConfigureAwait(false);
+        }
+
+        Task DoEncodeAsync(Stream innerStream)
+        {
+            try
+            {
+                // TODO: Are synchronous IO writes OK? We avoid reads.
```

Comment:
If it's a memory stream yes, otherwise no, since it will result in IO calls potentially blocking ThreadPool threads. We make sure to use `ChunkedMemoryStream` here as well.

Maybe better to do it in a separate PR though to test it out properly like we did in #1316.

Reply:
Will deal with it separately since I just moved the existing code to a new location.